### ISSUE TITLE
Onnx export of image_encoder

### ISSUE THREAD

**Issue by mgrewe** [heyoeyo/muggled_sam#7]
Dear all

once again I'm facing an issue when trying to export @heyoeyo's implementation of image_encoder to onnx. I'm using tracing-based exporter, since I wasn't successful with TorchScript and Dynamo. Seems like the issue lies in https://github.com/heyoeyo/muggled_sam/blob/3502a328cca4352e024487dd68d2551dd8a83b6a/lib/v2_sam/image_encoder_model.py#L78 

Error is
```
Exported graph: graph(%input : BFloat16(1, 3, 512, 512, strides=[3, 1, 1536, 3], requires_grad=0, device=cuda:0),
      %patch_embed.proj.weight : BFloat16(96, 3, 7, 7, strides=[147, 49, 7, 1], requires_grad=1, device=cuda:0),
      %patch_embed.proj.bias : BFloat16(96, strides=[1], requires_grad=1, device=cuda:0)):
  %output : BFloat16(1, 96, 128, 128, strides=[1572864, 16384, 128, 1], requires_grad=0, device=cuda:0) = onnx::Conv[dilations=[1, 1], group=1, kernel_shape=[7, 7], pads=[3, 3, 3, 3], strides=[4, 4], onnx_name="/patch_embed/proj/Conv"](%input, %patch_embed.proj.weight, %patch_embed.proj.bias), scope: lib.v2_sam.image_encoder_model.SAMV2ImageEncoder::/lib.v2_sam.components.imgenc_components.HalfStepPatchEmbed::patch_embed/torch.nn.modules.conv.Conv2d::proj # /usr/local/lib/python3.10/dist-packages/torch/nn/modules/conv.py:454:0
  return (%output)

/usr/local/lib/python3.10/dist-packages/torch/onnx/utils.py:1738: UserWarning: The exported ONNX model failed ONNX shape inference. The model will not be executable by the ONNX Runtime. If this is unintended and you believe there is a bug, please report an issue at https://github.com/pytorch/pytorch/issues. Error reported by strict ONNX shape inference: [ShapeInferenceError] (op_type:Conv, node name: /patch_embed/proj/Conv): X typestr: T, has unsupported type: tensor(bfloat16) (Triggered internally at ../torch/csrc/jit/serialization/export.cpp:1469.)
```

The original encoder module seems to export fine, but is quite different from this implementation and thus doesn't give me any hint. I know, onnx export is not the intention of this repo, but maybe someone has an idea what might be going on.

Cheers

**Comment by heyoeyo** [comment]
Hi @mgrewe, I've also been running into a bunch of issues with onnx exporting! The error messages tend to confuse me quite a bit.

> The original encoder module seems to export fine

The [original patch embedding model](https://github.com/facebookresearch/sam2/blob/c2ec8e14a185632b0a5d8b161928ceb50197eddc/sam2/modeling/backbones/utils.py#L91-L95) is more or less identical to the [muggled SAM version](https://github.com/heyoeyo/muggled_sam/blob/3502a328cca4352e024487dd68d2551dd8a83b6a/lib/v2_sam/components/imgenc_components.py#L75-L86) at least. I wonder if that means the errors are occurring at some other point and only breaking exporting once it gets to the patch embedding model... I know that the caching inside muggled SAM can cause some problems with exports at least.

For the error you posted, it mentions `unsupported type: tensor(bfloat16)` which is something I've run into as well. I'm not sure how to properly enable bfloat16 support with onnx, but it does seem like 'regular' float16 (or float32) works. So if your use case supports float16, that might be a way to get the export working.

Here's some code that seems to export the v2 image encoder without major errors:

<details>

<summary>Onnx export SAMv2 image encoder</summary>

```python
class SAMV2OnnxImageEncoder(nn.Module):
    """Wrapper around SAMv2 image encoder, used to help with onnx exporting"""

    def __init__(self, sam_model: SAMV2Model, image_size: int = 1024):

        # Inherit from parent
        super().__init__()

        # Store image encoder for use in forward pass
        self.model: SAMV2ImageEncoder = sam_model.image_encoder

        # Figure out compatible image sizing
        tiling_constraint = self.model.get_image_tiling_size_constraint()
        constrained_size = int(np.ceil(image_size / tiling_constraint) * tiling_constraint)
        self._scale_hw = (constrained_size, constrained_size)

        # Set up helper variable to keep track of model device & data type
        self.register_buffer("device_info", torch.empty((1)), persistent=False)

    def forward(self, image_rgb_tensor_bchw: torch.Tensor) -> torch.Tensor:

        # Scale input to target sizing & apply mean/scaling offsets
        scaled_img = nn.functional.interpolate(
            image_rgb_tensor_bchw.to(self.device_info),
            size=self._scale_hw,
            mode="bilinear",
            align_corners=False,
        )
        scaled_img = (scaled_img - self.model.mean_rgb) * self.model.stdev_scale_rgb
        image_encoding = self.model(scaled_img)

        return image_encoding

    def make_example_input(self) -> torch.Tensor:
        """Helper used to make example input 'args' for onnx export"""

        example_param = next(self.parameters())
        device, dtype = example_param.device, example_param.dtype
        return torch.randn((1, 3, *self._scale_hw), device=device, dtype=dtype)


# Settings
model_path = "/path/to/model.pt"
device = "cpu"
dtype = torch.float16
image_size = 512

# Create 'onnx-able' version of the image encoder for exporting
_, full_sam_model = make_samv2_from_original_state_dict(model_path)
imgenc_model = SAMV2OnnxImageEncoder(full_sam_model, image_size)
imgenc_model.to(device, dtype)

# Export model 'in memory' to avoid disk IO while testing
print("", "Exporting image encoder...", sep="\n", flush=True)
file_in_memory_imgenc = io.BytesIO()
torch.onnx.export(
    f=file_in_memory_imgenc,
    model=imgenc_model,
    args=tuple([imgenc_model.make_example_input()]),
    input_names=["input"],
    output_names=["output"],
    export_params=True,
    do_constant_folding=True,
    opset_version=14,
    dynamic_axes={
        "input": {0: "batch", 2: "height", 3: "width"},
        "output": {0: "batch", 2: "height", 3: "width"},
    },
)

# Confirm export result is ok
file_in_memory_imgenc.seek(0)
imgenc_onnx_model = onnx.load(file_in_memory_imgenc)
onnx.checker.check_model(imgenc_onnx_model)

# Save to disk
save_path = "/path/to/model.onnx"
enable_save = False
if enable_save:
    file_in_memory_imgenc.seek(0)
    with open(save_path, "wb") as file:
        file.write(file_in_memory_imgenc.getvalue())
    print("", "Saved onnx model:", save_path, sep="\n")
```

</details>


And here's code that can test the onnx model on an image:

<details>

<summary>Test onnx export</summary>

```python
import cv2
import numpy as np
import onnx
import onnxruntime

# Test image
image_path = "/path/to/image.jpg"
onnx_path = "/path/to/model.onnx"

# Make sure we can load the onnx model
# -> Can skip loading from a file here if this code runs after the
#    exporting code (i.e. directly use 'imgenc_onnx_model' from export)
#    this helps avoid lots of reading/writing to disk while testing
imgenc_onnx_model = onnx.load(onnx_path)
ort_session = onnxruntime.InferenceSession(
  imgenc_onnx_model.SerializeToString(),
  providers=["CPUExecutionProvider"],
)

# Load image and convert to RGB with batch/channels/height/width dimension ordering
# -> BGR-to-RGB could be built in to model if needed
# -> Instead assuming model will be used elsewhere, where RGB is already default format
img_bgr = cv2.imread(image_path)
img_rgb_bhwc = cv2.cvtColor(img_bgr, cv2.COLOR_BGR2RGB)
img_rgb_bchw = img_rgb_bhwc.swapaxes(0, 2)[None, :]

# Cast input to type expected by model
is_float16 = "float16" in ort_session.get_inputs()[0].type
img_rgb_bchw = img_rgb_bchw.astype(np.float16 if is_float16 else np.float32)

# Run the onnx model on real data to test it
onnx_output = ort_session.run(None, {"input": img_rgb_bchw})
print("", "Onnx model run results:", sep="\n")
print("  Num outputs:", len(onnx_output))
print("  Data type:", onnx_output[0].dtype)
print("  Example output shape:", onnx_output[0].shape)
```

</details>

That being said, this gives loads of warnings about needing to hard-code some of the boolean checks inside the model (which breaks support for changing input image sizing I believe...?), which I haven't figured out a solution for.

**Comment by mgrewe** [comment]
Hi @heyoeyo 

thanks for your answer. I think even if orig and this implementation are semantically equivalent, it might still fail in one case while working in the other - e.g., depending on which ops where used, and orig encoder exports fine with bfloat16. Have you compared results of float16 vs. bfloat16, did you experience significant loss of accuracy?

> I know that the caching inside muggled SAM can cause some problems with exports at least.

What kind of caching are you referring to? The memory bank or other module parameters? I haven't gotten very far yet, but I was thinking about arranging state entirely in tensors, esp. prompt_{memory*|object*} and previous_{memory*|object*}. Ideally, I would like to have computations entirely in nn.Modules and only maintain buffer memory from outside.

By the way: is there any empty placeholder (like `no_mem_embed`) that could be inserted into memory to maintain static tensor size during memory fusion, e.g., when tracking has just started and there are fewer previous frames than `max_memory_history`?

**Comment by mgrewe** [comment]
Just tried to comment-out everything except `patch_tokens_bhwc = self.patch_embed(image_tensor_bchw)` from SAMV2ImageEncoder's forward method. Error is still the same. Weird since `HalfStepPatchEmbed` only has a `nn.Conv2d` layer :shrug:

**Comment by heyoeyo** [comment]
> Have you compared results of float16 vs. bfloat16, did you experience significant loss of accuracy?

I haven't check it out too carefully for SAMv2, but with another model I found float16 is generally more 'accurate' (at least in terms of matching the float32 result) but can be prone to generating `NaN` or `inf` results in some cases (some more discussion of that can be found in the [dinov2 issues](https://github.com/facebookresearch/dinov2/issues/380#issuecomment-1971450039) board).
The muggled sam scripts use a [helper function](https://github.com/heyoeyo/muggled_sam/blob/3502a328cca4352e024487dd68d2551dd8a83b6a/lib/demo_helpers/misc.py#L58) that has a default arg of `prefer_bfloat16=True`. If you switch that to `False`, the scripts will all use float16 instead, so that's a quick way to check for differences. Here's a comparison flipping between bf16/f16/f32 two (showing colormapped mask logits):

![bf16_f16_f32](https://github.com/user-attachments/assets/bf6b4a81-8c4c-437b-93fd-e7baeddc2068)
There are noticeable differences between f16 & bf16, but it's not clear if one is better than the other.

> What kind of caching are you referring to?

For the image encoder, the caching shows up when computing the [position encoding](https://github.com/heyoeyo/muggled_sam/blob/3502a328cca4352e024487dd68d2551dd8a83b6a/lib/v2_sam/image_encoder_model.py#L126) for the patch embedding. The onnx exporter doesn't like the [boolean checks](https://github.com/heyoeyo/muggled_sam/blob/3502a328cca4352e024487dd68d2551dd8a83b6a/lib/v2_sam/components/imgenc_components.py#L154) used to check for cache mismatching, so it can cause problems, and it seems to flat-out fail if the cache exists (i.e. if you run the model once before exporting).
There are a number of other places where caching is used which you can find by searching for "persistent=False" since that's used to stop the cached tensors from being included in the model weights.

> I was thinking about arranging state fully as tensors

That's a good idea. I'm still working through updating the memory steps to properly include the changes introduced by the v2.1 models, but once that's finalized I'll probably make a similar change to taking tensors as inputs.

>  is there any empty placeholder that could be inserted into memory to maintain static tensor size during memory fusion

I don't think there's any learned embedding for that, but repeating the prompt memory to fill up all previous frame slots should probably work ok. Filling the memory with zeros also seems to work (may vary by video though!). A quick way to check this is to modify the [initial state](https://github.com/heyoeyo/muggled_sam/blob/3502a328cca4352e024487dd68d2551dd8a83b6a/simple_examples/video_segmentation.py#L55-L58) of the video_segmentation.py example script to something like:
```python
prev_mems = deque([init_mem] * 6, maxlen=6)
prev_ptrs = deque([init_ptr] * 15, maxlen=15)
```
This doesn't seem to (qualitatively) change the results for me. Using `[init_mem * 0.0] * 6` as the filler value also works.

> Just tried to comment-out everything except patch_tokens_bhwc = self.patch_embed(image_tensor_bchw)

Super weird! My experience with the onnx exporting has mostly been dealing with surprising errors as well. I'm guessing there's some set of config parameters for onnx that would resolve the issue, but I haven't looked into it enough to figure out yet.

**Comment by heyoeyo** [comment]
Actually one thing that might be worth doing is putting the original encoder onnx model into [netron](https://netron.app/) to view the structure (at least up until the patch embedding part). Here's what the muggled sam version (from the code above) looks like:

![onnx_netron](https://github.com/user-attachments/assets/57bde55b-750e-4faa-8d5e-ed5bacd07cc6)

The highlighted `Conv` block is where the exporting error occurs. Interestingly, though the bfloat16 version of muggled sam doesn't work, it can still be exported/saved and viewed in the netron app, and looks the same (but the initial cast is to bfloat16). So maybe the original encoder does something different to avoid the error?

**Comment by heyoeyo** [comment]
> Do you see a way of expressing this without conv op?

It could be done using a linear layer repeatedly applied to each of the patches, or otherwise as a single linear layer applied once to a matrix version of the image, re-arranged in a special way...
However, I think the extra work of shuffling the data around to make it work with linear layers would add a lot of overhead. The conversion you did (to float16, then back to bfloat16) is probably the easiest/fastest solution.

> However, next op missing bfloat16 support is MaxPool in HieraModel

The max pooling steps are super simple, it's just looking at 2x2 collections of values and keeping only the maximum value (out of the 4 values in the 2x2 grid). To see what it's doing, you can try running:
```python
import torch
data = torch.arange(16).reshape(1,1,4,4)
print(data)
# [[ 0,  1,  2,  3],
#  [ 4,  5,  6,  7],
#  [ 8,  9, 10, 11],
#  [12, 13, 14, 15]]

pool = torch.nn.MaxPool2d(2)
print(pool(data))
# [[ 5,  7],
#  [13, 15]]
```
So it's just taking the max value from the first 2x2 block (0, 1, 4, 5 -> max is 5), then the max of the next 2x2 block (2, 3, 6, 7 -> max is 7) and so on.

This is probably something that could be easily implemented manually, but indexing out the 2x2 blocks and then re-building the smaller resulting tensor would involve a lot of data shuffling/overhead. It may again just be simpler & faster to do the conversion to regular float16 (or float32 if stability is a concern) for the [maxpool step](https://github.com/heyoeyo/muggled_sam/blob/3502a328cca4352e024487dd68d2551dd8a83b6a/lib/v2_sam/components/hiera_blocks.py#L395) and then convert the result back to bfloat16 afterwards.
